src/vscode-services/window.ts

```diff
@@ -1,27 +1,30 @@
-import { CancellationTokenSource } from 'vs/base/common/cancellation'
 import Severity from 'vs/base/common/severity'
 import { URI } from 'vs/base/common/uri'
 import type * as vscode from 'vscode'
 import { Event } from 'vs/base/common/event'
+import * as extHostTypes from 'vs/workbench/api/common/extHostTypes'
+import workspace from './workspace'
+import { extension, getExtHostServices } from './extHost'
 import { Services } from '../services'
 import { unsupported } from '../tools'
 
-function showMessage<T extends vscode.MessageOptions | string | vscode.MessageItem> (type: Severity, message: string, ...rest: T[]): Thenable<T | undefined> {
-  const { window } = Services.get()
-  if (window == null) {
-    return Promise.resolve(undefined)
-  }
-  return window.showMessage(type, message, ...rest)
-}
-
 const window: typeof vscode.window = {
-  showInformationMessage: showMessage.bind(undefined, Severity.Info),
-  showWarningMessage: showMessage.bind(undefined, Severity.Warning),
-  showErrorMessage: showMessage.bind(undefined, Severity.Error),
+  showInformationMessage (message: string, ...rest: Array<vscode.MessageOptions | string | vscode.MessageItem>) {
+    const { extHostMessageService } = getExtHostServices()
+    return extHostMessageService.showMessage(extension, Severity.Info, message, rest[0], <Array<string | vscode.MessageItem>>rest.slice(1))
```

Comment:
why not removing the options (`rest[0]` ?) from the `...rest` directly in the args ?

Reply:
this code comes from vscode itself. They love to have method with a lot of parameter possibilities

In this case, I guess it's to be able to cast the remaining